hilti/toolchain/src/compiler/driver.cc

```diff
@@ -593,289 +597,44 @@ Result<Nothing> Driver::addInput(const hilti::rt::filesystem::path& path) {
-        _saveIterationAST(unit, plugin, "Transformed AST", 0);
-
-        if ( logger().isEnabled(logging::debug::AstPrintTransformed) )
-            hilti::print(std::cout, *unit->moduleRef());
-
-        if ( logger().errors() )
-            return result::Error("aborting after errors");
-    }
-
-    return Nothing();
-}
-
-/**
- * Filters a set of units for those associated with a specified plugin
- * extension. For those matching, includes all their dependencies for the same
- * extension as well. One can choose to have only fully resolved units
- * considered.
- */
-static auto _unitsForPlugin(const std::vector<std::shared_ptr<Unit>>& units, const std::string& extension,
-                            bool include_resolved) {
-    auto cmp = [](const auto& u1, const auto& u2) { return u1->uniqueID() < u2->uniqueID(); };
-    std::set<std::shared_ptr<Unit>, decltype(cmp)> nunits(cmp);
-
-    for ( auto&& u : units ) {
-        if ( u->extension() == extension && (! u->isResolved() || include_resolved) ) {
-            nunits.insert(u);
-
-            for ( const auto& d_ : u->dependencies(true) ) {
-                auto d = d_.lock();
-                assert(d);
-                if ( d->extension() == extension && (! d->isResolved() || include_resolved) )
-                    nunits.insert(d);
-            }
-        }
-    }
-
-    std::vector<std::shared_ptr<Unit>> nunits_vec;
-    nunits_vec.reserve(nunits.size());
-    for ( auto&& u : nunits )
-        nunits_vec.push_back(u);
-
-    return nunits_vec;
-}
-
-Result<Nothing> Driver::_resolveUnits() {
-    if ( _stage != Stage::INITIALIZED )
-        logger().internalError("unexpected driver stage in compileUnits()");
-
-    int round = 0;
-
-    auto plugin = plugin::registry().plugins().begin();
-    while ( plugin != plugin::registry().plugins().end() ) {
-        // Get remaining units that are relevant for the current plugin. Note
-        // that the list of pending units may change during this loop if more
-        // input files get added. If that happens, we will process any new ones
-        // that are not associated with plugins that we have already finish
-        // with.
-        if ( auto units = _unitsForPlugin(_pending_units, plugin->extension, false); units.size() ) {
-            if ( auto rc = _resolveUnitsWithPlugin(*plugin, units, round); ! rc )
-                return rc;
-        }
-        else {
-            // All done, switch to next plugin, but first perform any pending
-            // transformations.
-            auto all_units = _unitsForPlugin(_pending_units, plugin->extension, true);
-            if ( auto rc = _transformUnitsWithPlugin(*plugin, all_units); ! rc )
-                return rc;
-
-            ++plugin;
-
-            context()->dumpUnitCache(logging::debug::AstCache);
-        }
-    }
+Result<Nothing> Driver::addInput(const declaration::module::UID& uid) {
```

Comment:
We always copy below, we should probably instead take a value instead of a ref here and `move` below.